src/addons/pspassthrough.cpp

```diff
@@ -0,0 +1,167 @@
+          }
+      case PS4State::sending_nonce:
+          if ( !awaiting_cb ) {
+            uint8_t nonce_buffer[64];
+            nonce_buffer[0] = PS4AuthReport::PS4_GET_SIGNATURE_NONCE;
+            nonce_buffer[1] = PS4Data::getInstance().nonce_id;    // nonce_id
+            nonce_buffer[2] = send_nonce_part; // next_part
+            memset(&nonce_buffer[3], 0, 61); // zero rest of memory
+            uint8_t* buf = report_buffer;
+            uint16_t len = sizeof(nonce_buffer);
+            memcpy(buf, nonce_buffer, len);
+            send_nonce_part++; // Nonce Part is reset during callback
+            awaiting_cb = true;
+            tuh_hid_get_report(ps_dev_addr, ps_instance, PS4AuthReport::PS4_GET_SIGNATURE_NONCE, HID_REPORT_TYPE_FEATURE, buf, len);
+          }
+        break;
+    };
+}
+
+void PSPassthroughAddon::mount(uint8_t dev_addr, uint8_t instance, uint8_t const* desc_report, uint16_t desc_len) {
+    ps_dev_addr = dev_addr;
+    ps_instance = instance;
+
+    // Reset as soon as its connected
+    memset(report_buffer, 0, sizeof(report_buffer));
+    report_buffer[0] = PS4AuthReport::PS4_DEFINITION;
+    uint8_t* buf = report_buffer;
+    uint16_t len = 48;
+    awaiting_cb = true;
+    tuh_hid_get_report(ps_dev_addr, ps_instance, PS4AuthReport::PS4_DEFINITION, HID_REPORT_TYPE_FEATURE, buf, len);
+}
+
+void PSPassthroughAddon::unmount(uint8_t dev_addr) {
+    nonce_page = 0; // no nonce yet
+    send_nonce_part = 0; // which part of the nonce are we getting from send?
+    awaiting_cb = false; // did we receive the sign state yet
+    passthrough_state = PS4State::no_nonce;
+}
+
+void PSPassthroughAddon::set_report_complete(uint8_t dev_addr, uint8_t instance, uint8_t report_id, uint8_t report_type, uint16_t len) {
+    switch(report_id) {
+        case PS4AuthReport::PS4_SET_AUTH_PAYLOAD:
+            if (nonce_page == 5) {
+                nonce_page = 0;
+                passthrough_state = PS4State::signed_nonce_ready;
+            }
+            break;
+        default:
+            break;
+    };
+    awaiting_cb = false;
+}
+
+void PSPassthroughAddon::get_report_complete(uint8_t dev_addr, uint8_t instance, uint8_t report_id, uint8_t report_type, uint16_t len) {
+    switch(report_id) {
+        case PS4AuthReport::PS4_DEFINITION:
+            break;
+        case PS4AuthReport::PS4_RESET_AUTH:
+            if ( PS4Data::getInstance().ps4State == PS4State::nonce_ready)
+                passthrough_state = PS4State::receiving_nonce;
+            break;
+        case PS4AuthReport::PS4_GET_SIGNING_STATE:
+            if (report_buffer[2] == 0)
+                passthrough_state = PS4State::sending_nonce;
+            break;
+        case PS4AuthReport::PS4_GET_SIGNATURE_NONCE:
+            memcpy(&PS4Data::getInstance().ps4_auth_buffer[(send_nonce_part-1)*56], &report_buffer[4], 56);
+            if (send_nonce_part == 19) { // 0 = ready, 16 = not ready
```

Comment:
This coment belongs to the case above this one.

Reply:
Oh yeah, definitely had some copy+paste bits in there.

Reply:
I also discovered that 1 = could not sign in signing state, didn't know that error code.

Reply:
> I also discovered that 1 = could not sign in signing state, didn't know that error code.

Oh, any idea when this can happen? I think it will be good to add this to psdevwiki, as well as info about 0x03

Reply:
On the Mayflash chip, when I was working on the TInyUSB host code, I was asking for the signed nonce status before I had sent any valid nonce values. It kept sending 0x01.